### ISSUE TITLE
PDFsam basic 4 hangs at start

### ISSUE THREAD

**Issue by Speravir** [torakiki/pdfsam#357]
_TLDR:_ Go to the section below starting with “**Oooooh**”.

Hello, Windows 7 x64 here. After Oracle changed the license for their Java runtime I decided to switch to OpenJDK builds. In the moment I need both version 8 and version 12 (or 11) builds. They are also included in system PATH environment variable. This works fine for all other Java programs I have in use, but not for PDFsam, despite it has its own Java ressource shipped. I see the splash GIF, but this runs endlessly, at least for minutes until I started investigation.

First I started with `pdfsam.exe` where I noticed it hangs. The resulting command line in Process Hacker is:
```
"C:\Programs\PDF\pdfsam\runtime\bin\javaw.exe" -Xmx512m -Dprism.lcdtext=false -jar "C:\Programs\PDF\pdfsam\pdfsam-basic-4.0.3.jar"
```

Then I remembered there is a batch file in subfolder `bin`, so I started this from command line and got this:
```
C:\Programs\PDF\pdfsam\bin>pdfsam.bat
Exception in thread "WindowsNativeRunloopThread" java.lang.NoSuchMethodError: <i
nit>
        at com.sun.glass.ui.win.WinApplication.staticScreen_getScreens(Native Me
thod)
        at com.sun.glass.ui.Screen.initScreens(Screen.java:412)
        at com.sun.glass.ui.Application.lambda$run$1(Application.java:152)
        at com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
        at com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(WinApplication.j
ava:174)
        at java.base/java.lang.Thread.run(Unknown Source)
Exception in thread "JavaFX Application Thread" java.lang.NullPointerException
        at com.sun.prism.d3d.D3DPipeline.getAdapterOrdinal(D3DPipeline.java:205)

        at com.sun.javafx.tk.quantum.QuantumToolkit.assignScreensAdapters(Quantu
mToolkit.java:738)
        at com.sun.javafx.tk.quantum.QuantumToolkit.runToolkit(QuantumToolkit.ja
va:334)
        at com.sun.javafx.tk.quantum.QuantumToolkit.lambda$startup$10(QuantumToo
lkit.java:279)
        at com.sun.glass.ui.Application.lambda$run$1(Application.java:153)
        at com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
        at com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(WinApplication.j
ava:174)
        at java.base/java.lang.Thread.run(Unknown Source)
```

… and then it hangs until I shoot it down which leaves to these additional error message:
```
Exception in thread "Glass/Prism Shutdown Hook" java.lang.NullPointerException
        at com.sun.javafx.tk.quantum.QuantumToolkit.dispose(QuantumToolkit.java:
841)
        at com.sun.javafx.tk.quantum.QuantumToolkit$1.run(QuantumToolkit.java:25
2)
Batchvorgang abbrechen (J/N)?
```
(Yes, German here.)

**Oooooh**, while I wrote this I had an idea, and this works (!!!):
```
C:\Programs\PDF\pdfsam\bin>setlocal && set path=%CD%

C:\Programs\PDF\pdfsam\bin>echo %path%
C:\Programs\PDF\pdfsam\bin

C:\Programs\PDF\pdfsam\bin>pdfsam.bat
INFO  20:41:12.289 o.p.PdfsamApp[JavaFX-Launcher] Starting PDFsam
INFO  20:41:14.002 o.s.c.c.GlobalConfiguration[JavaFX Application Thread] Config
uring Sejda 3.2.75
INFO  20:41:15.311 o.p.PdfsamApp[JavaFX Application Thread] Gestarted in 3 secon
ds
```
So it seems, the reason is that PDFsam erroneously is trying to use one of the Java builds found in the system path or better: components of it. Perhaps the version 8 build which is in the path before the version 12 one?

Side remark: Remember in Windows `%path%` and `%PATH%` is the same – no letter case distinction here.

_Edit:_ Thinking about this a bit longer it must have be related to version 12: It worked with version 8 of Oracle Java which was in the system path, too. I added version 12 only, because one program does not work anymore with version 8.

**Comment by Speravir** [comment]
I cannot tell anything about the shell script, but I made a local change to the Windows batch as shipped with basic version 4.0.3, apparently [this one](https://github.com/torakiki/pdfsam/blob/0d883bc47dab3075e9c6e37fde888206bf953063/pdfsam-basic/src/assembly/resources/bin/pdfsam.bat).

I added the following to [line 66](https://github.com/torakiki/pdfsam/blob/0d883bc47dab3075e9c6e37fde888206bf953063/pdfsam-basic/src/assembly/resources/bin/pdfsam.bat#L66) (actually I added also 2 empty lines just for optical reasons):
```bat
set PATH=%RUNTIME%;%BASEDIR%
```
For WinNT this is enough, because everything is set local, but for Win9x (does someone use still use this?) we would need more steps:

Straight before line 66 above we should add:
```bat
set OLDPATH=%PATH%
```
… and before or after [line 96](https://github.com/torakiki/pdfsam/blob/0d883bc47dab3075e9c6e37fde888206bf953063/pdfsam-basic/src/assembly/resources/bin/pdfsam.bat#L96) we should add
```bat
set PATH=%OLDPATH%
```
(And adjust the comment in lines 94 and 95?)

**Comment by torakiki** [comment]
Thanks for investigating this. I already wrote to StackOverflow some time ago with no success, I now tried to open a bug in the jdk but I don't know if they will pick it up. I added your fix to the bat script but I think the .exe will still have this problem

**Comment by torakiki** [comment]
Here, a bug was opened on the jdk https://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8231015

**Comment by torakiki** [comment]
..and of course they closed/dismissed it without even reading it..

**Comment by Speravir** [comment]
Thanks nonetheless for your efforts.

> This is caused by a configuration error on the developers machine.

Quite interesting assertion. It does not work with the _default_ path here, but this is a configuration error. Funny they are.

**Comment by torakiki** [comment]
Would you be able to test one thing (we can't reproduce it so it's hard to test for us)? Add to the PATH variable the path to the bundled java runtime as first entry in the PATH and see if the pdfsam.exe starts.

**Comment by vicasong** [comment]
Any ideas? I'm having the same problem and I can't ask others to remove other jre.
